README.md
# Logstash Plugin

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

- Build gem
```sh
jruby -S gem build logstash-codec-avro_header.gemspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-codec-avro_header", :path => "/your/local/logstash-codec-avro_header"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'codec {avro_header {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-codec-avro_header.gemspec
```

- Install the plugin from a local gem
```sh
bin/logstash-plugin install /your/local/plugin/logstash-codec-avro_header.gem
```

or

- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.


CHANGELOG.md
## 3.2.4
  - Add support for decoding Kafka events with schema header

## 3.2.3
  - Update gemspec summary

## 3.2.2
  - Fix some documentation issues

## 3.2.0
 - Fixed an issue with the encoding that prevented certain fields from being serialized in a way compatible with the Kafka plugins

## 3.1.0
 - Introduce `tag_on_failure` option to tag events with `_avroparsefailure` instead of throwing an exception when decoding

## 3.0.0
 - breaking: Update to new Event API

## 2.0.4
 - Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash

## 2.0.3
 - New dependency requirements for logstash-core for the 5.0 release

## 2.0.0
 - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully,
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - Dependency on logstash-core update to 2.0


spec/codecs/avro_header_spec.rb
# encoding: utf-8
require 'logstash/devutils/rspec/spec_helper'
require 'avro'
require 'base64'
require 'logstash/codecs/avro_header'
require 'logstash/event'

describe LogStash::Codecs::AvroHeader do

  context "non binary data" do
    let (:avro_config) {{ 'schema_uri' => '
                        {"type": "record", "name": "Test",
                        "fields": [{"name": "foo", "type": ["null", "string"]},
                                   {"name": "bar", "type": "int"}]}' }}
    let (:test_event) {LogStash::Event.new({ "foo" => "hello", "bar" => 10 })}

    subject do
      allow_any_instance_of(LogStash::Codecs::AvroHeader).to \
      receive(:open_and_read).and_return(avro_config['schema_uri'])
      next LogStash::Codecs::AvroHeader.new(avro_config)
    end

    context "#decode" do
      it "should return an LogStash::Event from raw and base64 encoded avro data" do
        schema = Avro::Schema.parse(avro_config['schema_uri'])
        dw = Avro::IO::DatumWriter.new(schema)
        buffer = StringIO.new
        encoder = Avro::IO::BinaryEncoder.new(buffer)
        dw.write(test_event.to_hash, encoder)

        subject.decode(Base64.strict_encode64(buffer.string)) do |event|
          insist {event.is_a? LogStash::Event}
          insist {event.get("foo")} == test_event.get("foo")
          insist {event.get("bar")} == test_event.get("bar")
        end
        subject.decode(buffer.string) do |event|
          insist {event.is_a? LogStash::Event}
          insist {event.get("foo")} == test_event.get("foo")
          insist {event.get("bar")} == test_event.get("bar")
        end
      end

      it "should throw exception if decoding fails" do
        expect {subject.decode("not avro") {|_| }}.to raise_error NoMethodError
      end
    end

    context "#decode with tag_on_failure" do
      let (:avro_config) {super.merge("tag_on_failure" => true)}

      it "should tag event on failure" do
        subject.decode("not avro") do |event|
          insist {event.is_a? LogStash::Event}
          insist {event.get("tags")} == ["_avroparsefailure"]
        end
      end
    end

    context "#decode binary data with header" do
      let (:avro_config) {{ 'header_length' => 10,
                            'schema_uri' => '
                          {"type": "record", "name": "Test",
                          "fields": [{"name": "foo", "type": ["null", "string"]},
                                     {"name": "bar", "type": "int"}]}' }}

      it "should return an LogStash::Event from raw and base64 encoded avro data with a header" do
        schema = Avro::Schema.parse(avro_config['schema_uri'])
        dw = Avro::IO::DatumWriter.new(schema)
        buffer = StringIO.new
        header = [0xC3, 0x01, 0xDEADBEEF00000001].pack('CCQ<')
        buffer.write(header)

        encoder = Avro::IO::BinaryEncoder.new(buffer)
        dw.write(test_event.to_hash, encoder)

        subject.decode(Base64.strict_encode64(buffer.string)) do |event|
          insist {event.is_a? LogStash::Event}
          insist {event.get("foo")} == test_event.get("foo")
          insist {event.get("bar")} == test_event.get("bar")
        end
        subject.decode(buffer.string) do |event|
          insist {event.is_a? LogStash::Event}
          insist {event.get("foo")} == test_event.get("foo")
          insist {event.get("bar")} == test_event.get("bar")
        end
      end

      it "should throw exception if decoding fails" do
        expect {subject.decode("not avro") {|_| }}.to raise_error NoMethodError
      end
    end

    context "#decode binary data with header and marker" do
      let (:avro_config) {{ 'header_length' => 10,
                            'header_marker' => [0xC3, 0x01],
                            'schema_uri' => '
                          {"type": "record", "name": "Test",
                          "fields": [{"name": "foo", "type": ["null", "string"]},
                                     {"name": "bar", "type": "int"}]}' }}

      it "should return an LogStash::Event when the marker matches" do
        schema = Avro::Schema.parse(avro_config['schema_uri'])
        dw = Avro::IO::DatumWriter.new(schema)
        buffer = StringIO.new
        header = [0xC3, 0x01, 0xDEADBEEF00000001].pack('CCQ<')
        buffer.write(header)

        encoder = Avro::IO::BinaryEncoder.new(buffer)
        dw.write(test_event.to_hash, encoder)

        subject.decode(Base64.strict_encode64(buffer.string)) do |event|
          insist {event.is_a? LogStash::Event}
          insist {event.get("foo")} == test_event.get("foo")
          insist {event.get("bar")} == test_event.get("bar")
        end
      end

      it "should return an LogStash::Event when the marker doesn't match" do
        schema = Avro::Schema.parse(avro_config['schema_uri'])
        dw = Avro::IO::DatumWriter.new(schema)
        buffer = StringIO.new
        encoder = Avro::IO::BinaryEncoder.new(buffer)
        dw.write(test_event.to_hash, encoder)

        subject.decode(Base64.strict_encode64(buffer.string)) do |event|
          insist {event.is_a? LogStash::Event}
          insist {event.get("foo")} == test_event.get("foo")
          insist {event.get("bar")} == test_event.get("bar")
        end
      end

      it "should throw exception if decoding fails" do
        expect {subject.decode("not avro") {|_| }}.to raise_error NoMethodError
      end
    end


    context "#encode" do
      it "should return avro data from a LogStash::Event" do
        got_event = false
        subject.on_event do |event, data|
          schema = Avro::Schema.parse(avro_config['schema_uri'])
          datum = StringIO.new(Base64.strict_decode64(data))
          decoder = Avro::IO::BinaryDecoder.new(datum)
          datum_reader = Avro::IO::DatumReader.new(schema)
          record = datum_reader.read(decoder)

          insist {record["foo"]} == test_event.get("foo")
          insist {record["bar"]} == test_event.get("bar")
          insist {event.is_a? LogStash::Event}
          got_event = true
        end
        subject.encode(test_event)
        insist {got_event}
      end

      context "binary data" do

        let (:avro_config) {{ 'schema_uri' => '{"namespace": "com.systems.test.data",
                      "type": "record",
                      "name": "TestRecord",
                      "fields": [
                        {"name": "name", "type": ["string", "null"]},
                        {"name": "longitude", "type": ["double", "null"]},
                        {"name": "latitude", "type": ["double", "null"]}
                      ]
                    }' }}
        let (:test_event) {LogStash::Event.new({ "name" => "foo", "longitude" => 21.01234.to_f, "latitude" => 111.0123.to_f })}

        subject do
          allow_any_instance_of(LogStash::Codecs::AvroHeader).to \
      receive(:open_and_read).and_return(avro_config['schema_uri'])
          next LogStash::Codecs::AvroHeader.new(avro_config)
        end

        it "should correctly encode binary data" do
          schema = Avro::Schema.parse(avro_config['schema_uri'])
          dw = Avro::IO::DatumWriter.new(schema)
          buffer = StringIO.new
          encoder = Avro::IO::BinaryEncoder.new(buffer)
          dw.write(test_event.to_hash, encoder)

          subject.decode(Base64.strict_encode64(buffer.string)) do |event|
            insist {event.is_a? LogStash::Event}
            insist {event.get("name")} == test_event.get("name")
            insist {event.get("longitude")} == test_event.get("longitude")
            insist {event.get("latitude")} == test_event.get("latitude")
          end
        end
      end
    end
  end
end


spec/codecs/logstash-codec-avro_header_spec.rb
# encoding: utf-8
require_relative '../spec_helper'
require "logstash/codecs/logstash-codec-avro_header"


lib/logstash/codecs/logstash-codec-avro_header.rb
# encoding: utf-8
require "logstash/codecs/base"
require "logstash/namespace"

# This  codec will append a string to the message field
# of an event, either in the decoding or encoding methods
#
# This is only intended to be used as an example.
#
# input {
#   stdin { codec =>  }
# }
#
# or
#
# output {
#   stdout { codec =>  }
# }
#
class LogStash::Codecs::LogstashCodecAvroHeader < LogStash::Codecs::Base

  # The codec name
  config_name "logstash-codec-avro_header"

  # Append a string to the message
  config :append, :validate => :string, :default => ', Hello World!'

  def register
    @lines = LogStash::Codecs::Line.new
    @lines.charset = "UTF-8"
  end # def register

  def decode(data)
    @lines.decode(data) do |line|
      replace = { "message" => line.get("message").to_s + @append }
      yield LogStash::Event.new(replace)
    end
  end # def decode

  # Encode a single event, this returns the raw data to be returned as a String
  def encode_sync(event)
    event.get("message").to_s + @append + NL
  end # def encode_sync

end # class LogStash::Codecs::LogstashCodecAvroHeader


lib/logstash/codecs/avro_header.rb
# encoding: utf-8
require "open-uri"
require "avro"
require "base64"
require "logstash/codecs/base"
require "logstash/event"
require "logstash/timestamp"
require "logstash/util"

# Read serialized Avro records as Logstash events
#
# This plugin is used to serialize Logstash events as
# Avro datums, as well as deserializing Avro datums into
# Logstash events.
#
# ==== Encoding
#
# This codec is for serializing individual Logstash events
# as Avro datums that are Avro binary blobs. It does not encode
# Logstash events into an Avro file.
#
#
# ==== Decoding
#
# This codec is for deserializing individual Avro records. It is not for reading
# Avro files. Avro files have a unique format that must be handled upon input.
#
#
# ==== Usage
# To read messages from Kafka input:
#
# [source,ruby]
# ----------------------------------
# input {
#   kafka {
#     codec => avro {
#         schema_uri => "/tmp/schema.avsc"
#     }
#   }
# }
# filter {
#   ...
# }
# output {
#   ...
# }
# ----------------------------------
#
# Avro messages may have a header before the binary data indicating the schema,
# as described in the https://avro.apache.org/docs/1.8.2/spec.html#single_object_encoding[the Avro
# spec].
#
# The length of the header depends on how many bytes are used for a marker and
# the https://avro.apache.org/docs/1.8.2/spec.html#schema_fingerprints[fingerprint
# algorithm]. The fingerprint might be 8 bytes for CRC-64-AVRO, 16 bytes for
# MD5, or 32 bytes for SHA-256.
#
# Another option is the
# https://docs.confluent.io/current/schema-registry/docs/intro.html[Confluent
# Schema Registry], which uses a server to assign ids to different versions of
# schemas and look them up. That's supported by other logstash plugins, e.g.
# https://github.com/revpoint/logstash-codec-avro_schema_registry.
#
# At a minimum, specify the header length, and the plugin will skip those bytes
# before passing the input to the Avro parser.
#
# [source,ruby]
# ----------------------------------
# input {
#   kafka {
#     codec => avro {
#         schema_uri => "/tmp/schema.avsc"
#         header_length => 10
#     }
#   }
# }
# ----------------------------------
#
# If you specify the header marker, the plugin will attempt to match those
# bytes at the beginning of the data. If they match, it will skip the header
# bytes, otherwise it will pass all the data to the Avro parser.
#
# [source,ruby]
# ----------------------------------
# input {
#   kafka {
#     codec => avro {
#         schema_uri => "/tmp/schema.avsc"
#         header_length => 10
#         # Marker bytes, specified as integer
#         header_marker => [195, 1] # 0xC3, 0x01
#     }
#   }
# }
# ----------------------------------
class LogStash::Codecs::AvroHeader < LogStash::Codecs::Base
  config_name "avro_header"

  # schema path to fetch the schema from.
  # This can be a 'http' or 'file' scheme URI
  # example:
  #
  # * http - `http://example.com/schema.avsc`
  # * file - `/path/to/schema.avsc`
  config :schema_uri, :validate => :string, :required => true

  # number of header bytes to skip before the Avro data
  config :header_length, :validate => :number, :default => 0

  # marker bytes at beginning of header, specified as integers
  config :header_marker, :validate => :number, :list => true, :default => []

  # tag events with `_avroparsefailure` when decode fails
  config :tag_on_failure, :validate => :boolean, :default => false

  def open_and_read(uri_string)
    open(uri_string).read
  end

  public
  def register
    @schema = Avro::Schema.parse(open_and_read(schema_uri))
  end

  public
  def decode(data)
    begin
      binary_data = Base64.strict_decode64(data)
    rescue
      binary_data = data
    end

    datum = StringIO.new(binary_data)
    if header_length > 0
      if binary_data.length < header_length
        @logger.error('message is too small to decode header')
        # Ignore header and try to parse as Avro
      else
        if header_marker and header_marker.length > 0
          marker_length = header_marker.length
          marker = datum.read(marker_length)
          marker_bytes = marker.unpack("C" * marker_length)
          if marker_bytes == header_marker
            hash_length = header_length - marker_length
            if hash_length > 0
              datum.read(hash_length)
              # TODO: look up schema using hash
            end
          else
            @logger.error('header marker mismatch')
            # Assume that there is no header and try to parse as Avro
            datum.rewind
          end
        else
          # No marker, just read header and ignore it
          datum.read(header_length)
        end
      end
    end
    decoder = Avro::IO::BinaryDecoder.new(datum)
    datum_reader = Avro::IO::DatumReader.new(@schema)
    yield LogStash::Event.new(datum_reader.read(decoder))
  rescue => e
    if tag_on_failure
      @logger.error("Avro parse error, original data now in message field", :error => e)
      yield LogStash::Event.new("message" => data, "tags" => ["_avroparsefailure"])
    else
      raise e
    end
  end

  public
  def encode(event)
    dw = Avro::IO::DatumWriter.new(@schema)
    buffer = StringIO.new
    encoder = Avro::IO::BinaryEncoder.new(buffer)
    dw.write(event.to_hash, encoder)
    @on_event.call(event, Base64.strict_encode64(buffer.string))
  end

end
